### PR TITLE
Document the NodeAddressType values

### PR DESCRIPTION
**Pull request by danwinship**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
There is not a lot of clear guidance about what the `NodeAddressType` values mean, leading to occasional confusion and debate about things like [whether globally routable IPv6 node IPs should be considered "InternalIP" or "ExternalIP"](https://github.com/kubernetes/kubernetes/pull/86918#discussion_r363997750).

This tries to clarify, more or less summarizing how the existing cloud providers behave. I guess it does not actually resolve the question from the PR linked above though, but that's partly because there's not an unambiguously correct answer there and we kind of have to decide what the right answer is; currently internal IPs are basically assumed to be _both_ private and directly-routed, and external IPs are basically assumed to be _both_ public and indirectly-routed, so there's no unambiguously correct answer for IPs which are public and directly-routed... (though I'd still argue for "internal" by analogy with the bare metal case).

**Which issue(s) this PR fixes**:
none

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
If we agree on something here we should probably update https://kubernetes.io/docs/concepts/architecture/nodes/#addresses too

/sig network
/sig cloud-provider
/priority backlog

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+	// NodeHostName identifies the node's canonical hostname
+	NodeHostName NodeAddressType = "Hostname"
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may or may not be visible to hosts outside the cluster. By default it is
+	// assumed that kube-apiserver can reach node internal IPs, though it is possible
+	// to configure clusters where this is not the case.
```
Comment (aojea) [inline pkg/apis/core/types.go:4154]: just curious, when is this not possible?

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+	//
+	// In cloud environments, a NodeInternalIP address generally corresponds to an
+	// "internal" or "private" IP address which was automatically assigned to the
+	// node. In bare-metal clusters, NodeInternalIP is simply the standard address
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4158]: Is it worth documenting that the address associated with the default route is usually preferred?

File: pkg/apis/core/types.go [hunk 3]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+	// globally routable unicast IPs.)
+	//
+	// External IPs may be assigned directly to an interface on the node, like a
+	// NodeInternalIP, or alternatively, packets sent to the external IP may be NAT'ed
+	// to an internal node IP rather than being delivered directly (making the IP less
+	// efficient for node-to-node traffic than a NodeInternalIP).
+	//
+	// In cloud environments, a NodeExternalIP address generally corresponds to an
+	// "external", "public", or "virtual" IP address, and may have been manually
+	// assigned to the node by an administrator. In bare-metal clusters,
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4174]: I don't think  "bare-metal clusters" is accurate since this applies for cluster on cloud providers without the cloud provider integration and bare-metal clusters using a custom external cloud provider. I would rephrase this to say "In clusters without the cloud provider enabled,"

File: pkg/apis/core/types.go (hunk 1 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4154]: Some people want their control plane and the nodes to be on isolated networks from each other. The simplest form of this is that the control plane can't reach the internal IPs but the nodes all have externalIPs that the control plane can reach and you pass `--kubelet-preferred-address-types ExternalIP` to `kube-apiserver`. The more complicated form of this is that you set of some sort of authenticated proxy between the control plane and the nodes, which I forget the details of.

File: pkg/apis/core/types.go (hunk 2 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4158]: If you're only going to have _one_ node address then it should be the one associated with the default route, but that's not necessarily true about InternalIPs in general... (Like, if we were going to document that anywhere, it might make more sense to document it on the `NodeStatus.Addresses` field, not on `NodeAddressType`?)

File: pkg/apis/core/types.go (hunk 3 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4174]: updated

File: pkg/apis/core/types.go [hunk 4]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies the node's canonical hostname
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4145]: "In general, this should match the name of the node but is not strictly required." 

^ I don't know if this the last part is actually true since I've never seen cases where the hostname doesn't match.

File: pkg/apis/core/types.go [hunk 5]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may or may not be visible to hosts outside the cluster. By default it is
+	// assumed that kube-apiserver can reach node internal IPs, though it is possible
+	// to configure clusters where this is not the case.
+	//
+	// In cloud environments, a NodeInternalIP address generally corresponds to an
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4156]: I'd like to avoid terms like "In cloud environments" since it's not clear what a "cloud environment" means and I think this part of the document should be true to all environment anyways.

File: pkg/apis/core/types.go [hunk 6]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+	// "internal" or "private" IP address which was automatically assigned to the
+	// node. In clusters with no cloud provider, NodeInternalIP is simply the standard
+	// address type, and does not necessarily imply that the IP is internal-only.
+	NodeInternalIP NodeAddressType = "InternalIP"
+
+	// NodeExternalIP identifies an IP address explicitly intended for use from
+	// outside the cluster; it is assumed to be more public or more persistent than a
+	// NodeInternalIP. (In public clouds, NodeExternalIP addresses are normally
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4164]: "NodeExternalPs can be globally routable unicast IPs."

File: pkg/apis/core/types.go [hunk 7]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+	// globally routable unicast IPs.)
+	//
+	// External IPs may be assigned directly to an interface on the node, like a
+	// NodeInternalIP, or alternatively, packets sent to the external IP may be NAT'ed
+	// to an internal node IP rather than being delivered directly (making the IP less
+	// efficient for node-to-node traffic than a NodeInternalIP).
+	//
+	// In cloud environments, a NodeExternalIP address generally corresponds to an
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4172]: Same here, I think the "In cloud environments" part is not needed since this should be generally true for most (if not all) environments.

**Review by andrewsykim — COMMENTED** [review]
Minor nits about terminology but overall LGTM, thanks @danwinship

File: pkg/apis/core/types.go (hunk 5 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4156]: This part (and the corresponding part below for ExternalIP) was kind of written for future CloudProvider implementors: "here's what the other CloudProviders are doing; you should probably do the same." Especially, I was using the words "internal"/"private" vs "external"/"public"/"virtual" to point out specific words that tend to be used in cloud APIs, and show how they usually map to the Kubernetes APIs. This really doesn't apply to the non-CloudProvider use case.

> and I think this part of the document should be true to all environment anyways

I tried to be cloud-vs-non-cloud agnostic in the preceding paragraphs, but in this one I was very explicitly trying to _not_ be agnostic about that. For purposes of node addresses, there's a pretty fundamental distinction between "clusters that use `--cloud-provider`" and "clusters that don't use `--cloud-provider`". Maybe talking about "cloud environments" isn't the right way to describe that (just like saying "bare metal" wasn't).

File: pkg/apis/core/types.go (hunk 6 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4164]: I was trying to be more specific/normative than that. Public clouds generally have globally routable IPs at their disposable, and generally have APIs for assigning them to nodes, and are generally used in contexts in which the users want some or all nodes to be reachable from across the internet. Private clouds come with a different set of assumptions. And all of the "classic" public CloudProviders (AWS, GCP, Azure) implement NodeExternalIP in the way described here, while none of the classic private CloudProviders do. So it seems to make sense to call this out as being an expected behavior specifically of "public cloud" CloudProviders.

File: pkg/apis/core/types.go (hunk 5 above)
Comment (andrewsykim) [reply pkg/apis/core/types.go:4156]: > Maybe talking about "cloud environments" isn't the right way to describe that (just like saying "bare metal" wasn't).

"Cloud provider" is just the most common way of setting these addresses today but this API is independent of it and should not really refer to it IMO.

File: pkg/apis/core/types.go (hunk 6 above)
Comment (andrewsykim) [reply pkg/apis/core/types.go:4164]: > And all of the "classic" public CloudProviders (AWS, GCP, Azure) implement NodeExternalIP in the way described here, while none of the classic private CloudProviders do.

On vSphere, a VM can have a publically routable IP external IP. It's not common but you can.

File: pkg/apis/core/types.go (hunk 5 above)
Comment (aojea) [reply pkg/apis/core/types.go:4156]: > For purposes of node addresses, there's a pretty fundamental distinction between "clusters that use `--cloud-provider`" and "clusters that don't use `--cloud-provider`"

That can be another angle,from the perspective of "who" sets the IPs, i.e., "... on environments where kubelet delegates the assignation of IP addresses, ....  "

File: pkg/apis/core/types.go [hunk 8]
```diff
@@ -4142,11 +4142,48 @@ type NodeAddressType string
...
+
+	// NodeExternalIP identifies an IP address explicitly intended for use from
+	// outside the cluster; it is assumed to be more public or more persistent than a
+	// NodeInternalIP. (In public clouds, NodeExternalIP addresses are normally
+	// globally routable unicast IPs.)
+	//
+	// External IPs may be assigned directly to an interface on the node, like a
+	// NodeInternalIP, or alternatively, packets sent to the external IP may be NAT'ed
+	// to an internal node IP rather than being delivered directly (making the IP less
+	// efficient for node-to-node traffic than a NodeInternalIP).
```
Comment (aojea) [inline pkg/apis/core/types.go:4170]: there are a lot of implementation details here, can we make it more abstract?
"External IPs are not necessarily assigned to nodes, unlike InternalPs, making them less efficient for node-to-node traffic than a NodeInternalIP."

File: pkg/apis/core/types.go (hunk 5 above)
Comment (danwinship) [reply pkg/apis/core/types.go:4156]: > "Cloud provider" is just the most common way of setting these addresses today but this API is independent of it and should not really refer to it IMO.

Hm... ok, I guess I'm mixing up the documentation of `NodeAddressType` with the documentation of `Node.Status.Addresses`. _In theory_, someone could use `NodeAddressType` for something else, but currently, it's only used for `Node.Status.Addresses`, and in that context, it is not independent of cloud providers. (Kubelet requires that either kubelet or an external cloud provider owns `Node.Status.Addresses`.)

File: pkg/apis/core/types.go (hunk 4 above)
Comment (anguslees) [reply pkg/apis/core/types.go:4145]: This is a big terminology mess in k8s, but: Node name != `os.Hostname()` (aka `uname -n`) != DNS entry
It gets worse, because (eg) `kubelet --hostname-override` actually sets the node name, not the hostname. And other examples.

On my on-prem cluster, for example: None of the hosts have entries in DNS, and I use /etc/machine-id as the node name (ie: `kubelet --hostname-override=%m` in systemd config).  `uname -n` reports "localhost" (CoreOS/flatcar out-of-the-box default) on all of them.  Kubernetes works fine in this setup, and honestly I think it's superior to trying to align DNS primary keys for no reason.  Other common examples include OpenStack, where there is (was?) commonly no DNS entry for hosts but the node-name == openstack instance name, and AWS with DHCP options, where the hostname (domain suffix) might not match the generated node name.  Basically the only place where node name == uname -n == DNS is GCE, so it's not surprising that k8s had this confused initially :stuck_out_tongue:

**Comment by danwinship** [comment]
Updated and simplified. Removed all discussion of cloud providers and replaced some of the discussion of semantics with vaguer handwaving about the semantics being not-clearly-defined (especially for external IPs).

**Comment by aojea** [comment]
/retest
/lgtm
_it is what it is_ :shrug: :)
/assign @thockin @dcbw @andrewsykim

File: pkg/apis/core/types.go (hunk 4 above)
Comment (aojea) [reply pkg/apis/core/types.go:4258]: I think that canonical hostname is ok, we are not going to solve a 30 years old problem ;)

Comment (anguslees) [reply pkg/apis/core/types.go:4258]: > I think that canonical hostname is ok

Where 'canonical hostname' means `os.Hostname()` or "DNS entry" (or `Node.metadata.name`) ?  Please pick one, because I don't think 'canonical hostname' is at all clear in k8s context.  I'm not asking you to solve anything except this comment :)

I suggest `os.Hostname()` (aka `uname -n`) - since we already have NodeInternalDNS/NodeExternalDNS for "DNS Entry", and we obviously already have the `Node` name elsewhere in the `Node` object.

Comment (aojea) [reply pkg/apis/core/types.go:4258]: is it clear in a typical unix system? :)

Comment (danwinship) [reply pkg/apis/core/types.go:4258]: Current behavior is:
- AWS: primary DNS name
- Azure: `node.Name`
- GCE: primary DNS name
- OpenStack: "hostname". not clear without further digging
- vSphere: `os.Hostname()`
- "bare metal": `os.Hostname()` or command-line override

I had thought it was more consistent than that. So yeah, this isn't a canonical anything.

Comment (anguslees) [reply pkg/apis/core/types.go]: Thanks for the survey!  So .. how should we document the current semantics of this? As "an opaque human readable string"?  Basically you can't rely on it being anything in particular, other than a value related to the node 'somehow'.

Maybe we just own up to the current state, and say: "NodeHostName is not a useful value. It can be _any_ of DNS name, os.Hostname or node.Name - and the expected semantics are undefined.  We recommend using NodeInternalDNS or node.Name instead."

Comment (aojea) [reply pkg/apis/core/types.go]: Maybe I'm not getting right the english meaning of canonical , but I was thinking that  "canonical name is what the k8s admin, cluster provider, or whoever sets the names on the nodes want to use for the primary name of the node, independently on how it gets that name ... in DNS the CNAME is a record and there is not rule too ...

**Review by thockin — COMMENTED** [review]
/lgtm
/approved
/hold

I am OK with these definitions, though on reading them I am not convinced this field is reliably useful for much of anything (at least, not across platforms) :)

If the other comments are just stale, feel free to nix the hold.

**Comment by danwinship** [comment]
I'll give @andrewsykim and @anguslees a chance to comment on whether they're OK with my latest changes.

Also @thockin you typoed `/approve` as `/approved`

> I am OK with these definitions, though on reading them I am not convinced this field is reliably useful for much of anything (at least, not across platforms) :)

Yeah, my original goal in this was to settle the question of whether globally-routable IPv6 addresses on AWS were "internal" or "external", and in the end it doesn't really answer that. (Though it does point in the direction that I was arguing by saying that "internal" is the default type and a node should always have an IP of that type, and "external" is defined mostly only as "more external than an internal IP". So from that point of view, since there is only one type of IPv6 address in AWS, it should be considered "internal", not "external".)

**Review by andrewsykim — COMMENTED** [review]
/lgtm

Thanks @danwinship!

File: pkg/apis/core/types.go [hunk 9]
```diff
@@ -4255,11 +4255,44 @@ type NodeAddressType string
 
 // These are valid values of node address type
 const (
-	NodeHostName    NodeAddressType = "Hostname"
-	NodeExternalIP  NodeAddressType = "ExternalIP"
-	NodeInternalIP  NodeAddressType = "InternalIP"
-	NodeExternalDNS NodeAddressType = "ExternalDNS"
+	// NodeHostName identifies a name of the node. Although every node can be assumed
+	// to have a NodeAddress of this type, its exact syntax and semantics are not
+	// defined, and are not consistent between different clusters.
```
Comment (andrewsykim) [inline pkg/apis/core/types.go:4260]: Non-blocking comment: we probably want to mention here that this usually matches the hostname according to the OS, although it's not strictly required.

Comment (danwinship) [reply pkg/apis/core/types.go:4260]: Per https://github.com/kubernetes/kubernetes/pull/95768#discussion_r612656206 I don't think that's true

**Comment by danwinship** [comment]
/hold cancel
but @thockin you need to `/approve` still (you said `/approveD` before)

File: pkg/apis/core/types.go (hunk 9 above)
Comment (andrewsykim) [reply pkg/apis/core/types.go:4260]: I did say "usually" ;) "primary DNS name" and "hostname" is typically the same from my experience

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -4255,11 +4255,44 @@ type NodeAddressType string
...
+	NodeHostName NodeAddressType = "Hostname"
+
+	// NodeInternalIP identifies an IP address which is assigned to one of the node's
+	// network interfaces. Every node should have at least one address of this type.
+	//
+	// An internal IP is normally expected to be reachable from every other node, but
+	// may not be visible to hosts outside the cluster. By default it is assumed that
+	// kube-apiserver can reach node internal IPs, though it is possible to configure
+	// clusters where this is not the case.
```
Comment (sftim) [inline pkg/apis/core/types.go:4269]: I believe we're also asserting that any Pod can at least _unambigously identify_ an NodeInternalIP address as an internal IP, even if the Pod may be prevented from sending traffic to / from that internal Node IP.

Does that sound right?

For example, if I'm running as a Pod `foo` in some namespace in the cluster and I learn that 10.42.0.1 is a node IP address, that doesn't ever conflict with the IP address for a Service, another Pod, the API server, etc. I can recognise that a connection from 10.42.0.1 could only have originated from that node (maybe it is performing a TCP health check as I am executing on that particular node).

Comment (andrewsykim) [reply pkg/apis/core/types.go:4269]: >I can recognise that a connection from 10.42.0.1 could only have originated from that node (maybe it is performing a TCP health check as I am executing on that particular node).

I think this gets tricky given some edge scenarios regarding SNAT

**Comment by danwinship** [comment]
/reopen
/remove-lifecycle rotten

@thockin you had tried to approve this before but typoed `/approve`

**Comment by thockin** [comment]
/approve 

SORRY!!
